src/ERC1271.sol

```diff
@@ -1,50 +1,77 @@
+///
+/// @notice Abstract ERC-1271 implementation (based on Solady's one) with guards to handle the same
+///         signer being used on multiple accounts, using a nested approach based on EIP-712.
+///
+/// @dev To prevent the same signature from being validated on different accounts owned by the samer signer,
+///      we introduce an anti cross-account-replay layer: the original signed hash (OSH) is wrapped in a
+///      `CoinbaseSmartWalletMessage` (CSWM) struct, which is itself hashed using EIP-712.
+///
+///      When hashing the CSWM, the `domainSeparator` used has its `verifyingContract` field set to the address
+///      of the targeted account. This mechanism is coupling the signature of the CSWM hash with the account,
+///      effectively preventing it from being replayed on another account owned by the same signer.
+///
+///      See `replaySafeHash()` for the implementation details.
+///
 /// @author Coinbase (https://github.com/coinbase/smart-wallet)
+/// @author Solady (https://github.com/vectorized/solady/blob/main/src/accounts/ERC1271.sol)
 abstract contract ERC1271 {
-    /// @dev We use `bytes32 hash` rather than `bytes message`
-    /// In the EIP-712 context, `bytes message` would be useful for showing users a full message
-    /// they are signing in some wallet preview. But in this case, to prevent replay
-    /// across accounts, we are always dealing with nested messages, and so the
-    /// input should be a EIP-191 or EIP-712 output hash.
-    /// E.g. The input hash would be result of
+    /// @dev Precomputed `typeHash` used to produce the CSWM hash using EIP-712.
     ///
-    ///  keccak256("\x19\x01" || someDomainSeparator || hashStruct(someStruct))
+    ///      The OSH must either be:
+    ///         - An EIP-191 hash: keccak256("\x19Ethereum Signed Message:\n" || len(someMessage) || someMessage)
+    ///         - An EIP-712 hash: keccak256("\x19\x01" || someDomainSeparator || hashStruct(someStruct))
+    bytes32 private constant _MESSAGE_TYPEHASH = keccak256("CoinbaseSmartWalletMessage(bytes32 hash)");
+
+    /// @notice Validates the `signature` against the given `hash` after wrapping it in a CSWM. It allows the
+    ///         implementing account to be used as a signer.
     ///
-    ///  OR
+    /// @dev This implementation follows ERC-1271. See https://eips.ethereum.org/EIPS/eip-1271.
+    /// @dev IMPORTANT: The signature is validated against the CSWM hash.
     ///
-    ///  keccak256("\x19Ethereum Signed Message:\n" || len(someMessage) || someMessage),
+    /// @param hash      The original signed hash.
+    /// @param signature The signature of the CSWM hash to validate.
     ///
-    bytes32 private constant _MESSAGE_TYPEHASH = keccak256("CoinbaseSmartWalletMessage(bytes32 hash)");
-
-    /// @dev Validates the signature with ERC1271 return,
-    /// so that this account can also be used as a signer.
+    /// @return result `0x1626ba7e` if validation succeeded, else `0xffffffff`.
```

Comment:
Despite referencing EIP-1271 above, it might be clearer to define these "magic numbers" as constants above. If you opt to leave them as comments, then I think it's helpful to describe the derivation in the comment like: 

```suggestion
    /// @return result `0x1626ba7e` == bytes4(keccak256("isValidSignature(bytes32,bytes)") upon success, else `0xffffffff`
```